src/handlers/RF24StreamLogHandler.cpp

```diff
@@ -0,0 +1,217 @@
+
+RF24StreamLogHandler::RF24StreamLogHandler(Stream *stream)
+{
+   this->stream = stream;
+   this->logLevel = RF24LogLevel::INFO;
+}
+
+void RF24StreamLogHandler::log(uint8_t logLevel,
+                                   const __FlashStringHelper *vendorId,
+                                   const char *message,
+                                   va_list *args)
+{
+   if(logLevel > this->logLevel)
+   {
+      return;
+   }
+
+   appendTimestamp();
+   appendLogLevel(logLevel);
+   appendVendorId(vendorId);
+
+   // print formatted message
+   appendFormattedMessage(message, args);
+   stream->println("");
+}
+
+void RF24StreamLogHandler::log(uint8_t logLevel,
+                                   const __FlashStringHelper *vendorId,
+                                   const __FlashStringHelper *message,
+                                   va_list *args)
+{
+   if(logLevel > this->logLevel)
+   {
+      return;
+   }
+
+   appendTimestamp();
+   appendLogLevel(logLevel);
+   appendVendorId(vendorId);
+
+   // print formatted message
+   appendFormattedMessage(message, args);
+   stream->println("");
+}
+
+void RF24StreamLogHandler::setLogLevel(uint8_t logLevel)
+{
+   this->logLevel = logLevel;
+}
+
+void RF24StreamLogHandler::appendFormattedMessage(const char *format, va_list *args)
+{
+   for (; *format != 0; ++format)
+   {
+      if (*format == '%')
+      {
+         ++format;
+         appendFormat(*format, args);
+      }
+      else
+      {
+         stream->print(*format);
+      }
+   }
+}
+
+void RF24StreamLogHandler::appendFormattedMessage(const __FlashStringHelper *format, va_list *args)
+{
+   PGM_P p = reinterpret_cast<PGM_P>(format);
+   char c = pgm_read_byte(p++);
+   for (; c != 0; c = pgm_read_byte(p++))
+   {
+      if (c == '%')
+      {
+         c = pgm_read_byte(p++);
+         appendFormat(c, args);
+      }
+      else
+      {
+         stream->print(c);
+      }
+   }
+}
+
+void RF24StreamLogHandler::appendFormat(const char format, va_list *args)
+{
+   if (format == 's')
+   {
+      // print text from RAM
+      register char *s = (char *)va_arg(*args, int);
+      stream->print(s);
+
+      return;
+   }
+
+   if (format == 'S')
+   {
+      // print text from FLASH
+      register __FlashStringHelper *s = (__FlashStringHelper *)va_arg(*args, int);
+      stream->print(s);
+
+      return;
+   }
+
+   if (format == 'D' || format == 'F')
+   {
+      // print as double
+      stream->print(va_arg(*args, double));
+
+      return;
+   }
+
+   if (format == 'd' || format == 'i')
+   {
+      // print as integer
+      stream->print(va_arg(*args, int), DEC);
+
+      return;
+   }
+
+   stream->print(format);
+}
+
+void RF24StreamLogHandler::appendTimestamp()
+{
+   char c[12];
+   sprintf(c, "%10lu ", millis());
+   stream->print(c);
+   stream->print(" ");
+}
+
+void RF24StreamLogHandler::appendLogLevel(uint8_t logLevel)
+{
+   uint8_t logMainLevel = logLevel & 0b11111000;

```

Comment:
This is how the main log level (ERROR, WARN, INFO, DEBUG, TRACE) is calculated from the logLevel variable.